server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,710 @@
+          Set<String> storageGroupNamesNeedCreated = new HashSet<>();
+          for (String devicePath : failedDevices) {
+            PartialPath storageGroupNameNeedCreated =
+                MetaUtils.getStorageGroupPathByLevel(
+                    new PartialPath(devicePath), config.getDefaultStorageGroupLevel());
+            storageGroupNamesNeedCreated.add(storageGroupNameNeedCreated.getFullPath());
+          }
+          createNotExistedStorageGroup(storageGroupNamesNeedCreated);
+          // third try to hit cache
+          boolean thirdTryResult =
+              getDeviceToStorageGroupMap(
+                  devicePaths, deviceToStorageGroupMap, new ArrayList<>(), true);
+          if (!thirdTryResult) {
+            throw new StatementAnalyzeException("Failed to get Storage Group Map in three try.");
+          }
+        }
+      } catch (TException | MetadataException | IOException e) {
+        throw new StatementAnalyzeException(
+            "An error occurred when executing getDeviceToStorageGroup():" + e.getMessage());
+      }
+    }
+    return deviceToStorageGroupMap;
+  }
+
+  /** get all storage group from confignode and update storage group cache */
+  private void getAllStorageGroupFromConfigNode() throws IOException, TException {
+    try (ConfigNodeClient client =
+        configNodeClientManager.borrowClient(ConfigNodeInfo.partitionRegionId)) {
+      TStorageGroupSchemaResp storageGroupSchemaResp =
+          client.getMatchedStorageGroupSchemas(ROOT_PATH);
+      if (storageGroupSchemaResp.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        Set<String> storageGroupNames = storageGroupSchemaResp.getStorageGroupSchemaMap().keySet();
+        // update all storage group into cache
+        updateStorageCache(storageGroupNames);
+      }
+    }
+  }
+
+  /** create not existed storage group and update storage group cache */
+  private void createNotExistedStorageGroup(Set<String> storageGroupNamesNeedCreated)
+      throws IOException, TException {
+    try (ConfigNodeClient client =
+        configNodeClientManager.borrowClient(ConfigNodeInfo.partitionRegionId)) {
+      Set<String> successFullyCreatedStorageGroup = new HashSet<>();
+      for (String storageGroupName : storageGroupNamesNeedCreated) {
+        TStorageGroupSchema storageGroupSchema = new TStorageGroupSchema();
+        storageGroupSchema.setName(storageGroupName);
+        TSetStorageGroupReq req = new TSetStorageGroupReq(storageGroupSchema);
+        TSStatus tsStatus = client.setStorageGroup(req);
+        if (TSStatusCode.SUCCESS_STATUS.getStatusCode() == tsStatus.getCode()) {
+          successFullyCreatedStorageGroup.add(storageGroupName);
+        } else {
+          updateStorageCache(successFullyCreatedStorageGroup);
+          throw new RuntimeException(new IoTDBException(tsStatus.message, tsStatus.code));
+        }
+      }
+      updateStorageCache(storageGroupNamesNeedCreated);
```

Comment:
Would it be better if we log out all the StorageGroups that were created successfully or failed?

Reply:
Get it ~